File: L17AdobeMedium/Sudoku.cpp
#include <iostream>
#include <vector>
#include <set>
using namespace std;

class Solution
{
public:
    bool isValidSudoku(vector<vector<char> > &board)
    {
        // Three sets to track rows, columns, and subgrids
        set<char> vf[9];    // To track the numbers in each column
        set<char> hf[9];    // To track the numbers in each row
        set<char> sf[3][3]; // To track the numbers in each 3x3 subgrid

        // Loop through each cell in the board
        for (int i = 0; i < 9; i++)
        {
            for (int j = 0; j < 9; j++)
            {
                char num = board[i][j];

                if (num != '.')
                {
                    // Check if the number is already present in the row, column, or subgrid
                    if (hf[i].find(num) != hf[i].end() || vf[j].find(num) != vf[j].end() || sf[i / 3][j / 3].find(num) != sf[i / 3][j / 3].end())
                    {
                        return false; // If found, return false
                    }

                    // Insert the number into the respective row, column, and subgrid sets
                    hf[i].insert(num);
                    vf[j].insert(num);
                    sf[i / 3][j / 3].insert(num);
                }
            }
        }

        return true; // If no duplicates are found, the Sudoku is valid
    }
};

int main()
{
    Solution solution;

    // Define a sample Sudoku board
    vector<vector<char> > board = {
        {'5', '3', '.', '.', '7', '.', '.', '.', '.'},
        {'6', '.', '.', '1', '9', '5', '.', '.', '.'},
        {'.', '9', '8', '.', '.', '.', '.', '6', '.'},
        {'8', '.', '.', '.', '6', '.', '.', '.', '3'},
        {'4', '.', '.', '8', '.', '3', '.', '.', '1'},
        {'7', '.', '.', '.', '2', '.', '.', '.', '6'},
        {'.', '6', '.', '.', '.', '.', '2', '8', '.'},
        {'.', '.', '.', '4', '1', '9', '.', '.', '5'},
        {'.', '.', '.', '.', '8', '.', '.', '7', '9'}};

    cout << "Is the Sudoku valid? " << (solution.isValidSudoku(board) ? "Yes" : "No") << endl;
    return 0;
}

File: L17AdobeMedium/Sudoku.java

import java.util.HashSet;

public class Sudoku {

    public boolean isValidSudoku(char[][] board) {
        // Create sets to track seen numbers in rows, columns, and 3x3 subgrids
        HashSet<String> seen = new HashSet<>();

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                char num = board[i][j];
                if (num != '.') {
                    // Create unique keys for rows, columns, and subgrids
                    String rowKey = "row" + i + num;
                    String colKey = "col" + j + num;
                    String subgridKey = "subgrid" + (i / 3) + (j / 3) + num;

                    // If any of the keys already exist, the board is invalid
                    if (seen.contains(rowKey) || seen.contains(colKey) || seen.contains(subgridKey)) {
                        return false;
                    }

                    // Otherwise, add the keys to the set
                    seen.add(rowKey);
                    seen.add(colKey);
                    seen.add(subgridKey);
                }
            }
        }

        return true;
    }

    public static void main(String[] args) {
        Sudoku solution = new Sudoku();
        char[][] board = {
            {'5', '3', '.', '.', '7', '.', '.', '.', '.'},
            {'6', '.', '.', '1', '9', '5', '.', '.', '.'},
            {'.', '9', '8', '.', '.', '.', '.', '6', '.'},
            {'8', '.', '.', '.', '6', '.', '.', '.', '3'},
            {'4', '.', '.', '8', '.', '3', '.', '.', '1'},
            {'7', '.', '.', '.', '2', '.', '.', '.', '6'},
            {'.', '6', '.', '.', '.', '.', '2', '8', '.'},
            {'.', '.', '.', '4', '1', '9', '.', '.', '5'},
            {'.', '.', '.', '.', '8', '.', '.', '7', '9'}
        };

        System.out.println("Is the Sudoku valid? " + solution.isValidSudoku(board));
    }
}
